clippy_lints/src/redundant_locals.rs

```diff
@@ -82,6 +85,31 @@ impl<'tcx> LateLintPass<'tcx> for RedundantLocals {
     }
 }
 
+/// Checks if the enclosing body is a closure and if the given local is captured by value.
+///
+/// In those cases, the redefinition may be necessary to force a move:
+/// ```
+/// fn assert_static<T: 'static>(_: T) {}
+///
+/// let v = String::new();
+/// let closure = || {
+///   let v = v; // <- removing this redefinition makes `closure` no longer `'static`
+///   dbg!(&v);
+/// };
+/// assert_static(closure);
+/// ```
+fn is_closure_capture(cx: &LateContext<'_>, redefinition: HirId, root_variable: HirId) -> bool {
+    let body = cx.tcx.hir().enclosing_body_owner(redefinition);
+    if let DefKind::Closure = cx.tcx.def_kind(body) {
+        cx.tcx.closure_captures(body).iter().any(|c| {
+            matches!(c.info.capture_kind, UpvarCapture::ByValue)
+                && matches!(c.place.base, PlaceBase::Upvar(upvar) if upvar.var_path.hir_id == root_variable)
+        })
+    } else {
+        false
+    }
```

Comment:
good point, i'll change it later today

Reply:
Actually ended up using `tcx.is_closure_or_coroutine` (and added tests for coroutines) since that seems more correct, but the suggestion here to get rid of the `if` and simplify it to `&&` still applies so thanks :)